### PR TITLE
8294217: Assertion failure: parsing found no loops but there are some

### PR DESCRIPTION
**Pull request by rwestrel**
This was reported on 11 and is not reproducible with the current
jdk. The reason is that the PhaseIdealLoop invocation before EA was
changed from LoopOptsNone to LoopOptsMaxUnroll. In the absence of
loops, LoopOptsMaxUnroll exits earlier than LoopOptsNone. That wasn't
intended and this patch makes sure they behave the same. Once that's
changed, the crash reproduces with the current jdk.

The assert fires because PhaseIdealLoop::only_has_infinite_loops()
returns false even though the IR only has infinite loops. There's a
single loop nest and the inner most loop is an infinite loop. The
current logic only looks at loops that are direct children of the root
of the loop tree. It's not the first bug where
PhaseIdealLoop::only_has_infinite_loops() fails to catch an infinite
loop (8257574 was the previous one) and it's proving challenging to
have PhaseIdealLoop::only_has_infinite_loops() handle corner cases
robustly. I reworked PhaseIdealLoop::only_has_infinite_loops() once
more. This time it goes over all children of the root of the loop
tree, collects all controls for the loop and its inner loop. It then
checks whether any control is a branch out of the loop and if it is
whether it's not a NeverBranch.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294217](https://bugs.openjdk.org/browse/JDK-8294217): Assertion failure: parsing found no loops but there are some

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10904/head:pull/10904` \
`$ git checkout pull/10904`

Update a local copy of the PR: \
`$ git checkout pull/10904` \
`$ git pull https://git.openjdk.org/jdk pull/10904/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10904`

View PR using the GUI difftool: \
`$ git pr show -t 10904`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10904.diff">https://git.openjdk.org/jdk/pull/10904.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -4182,22 +4182,41 @@ bool PhaseIdealLoop::process_expensive_nodes() {
...
+          wq.push(c->in(j));
+        }
+      } else {
+        wq.push(c->in(0));
+      }
+    }
+    assert(wq.member(head), "");
+    for (uint i = 0; i < wq.size(); ++i) {
+      Node* c = wq.at(i);
+      if (c->isa_MultiBranch()) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:4209]: Can be changed to `is_MultiBranch()` as you do not need the casted type.

File: test/hotspot/jtreg/compiler/loopopts/TestInfiniteLoopNest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            l++;
+            while (l != 1) --l;
+            l = 9;
+        } while (l != 5);
+
+    }
+
+    public static void main(String[] p) throws Exception {
+        Thread thread = new Thread() {
+                public void run() {
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/loopopts/TestInfiniteLoopNest.java:55]: Indentation is wrong here.

File: src/hotspot/share/opto/loopnode.cpp [hunk 3]
```diff
@@ -4182,22 +4182,41 @@ bool PhaseIdealLoop::process_expensive_nodes() {
 #ifdef ASSERT
 bool PhaseIdealLoop::only_has_infinite_loops() {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:4183]: > This time it goes over all children of the root of the loop
tree, collects all controls for the loop and its inner loop. It then
checks whether any control is a branch out of the loop and if it is
whether it's not a NeverBranch.

Maybe you can add this summary as a comment here.

**Review by chhagedorn — APPROVED** [review]
That looks reasonable to me.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by rwestrel** [comment]
@chhagedorn thanks for reviewing! I pushed a new commit that should address your comments.
@TobiHartmann thanks for reviewing!

**Review by chhagedorn — APPROVED** [review]
Looks good, thanks for doing the updates!

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**